ArrayType must be represented by AnnotatedArrayType for un-inferred wildcard

Take this example, which is based on #419, and run a checker on it (tested with Regex and Nullness):

```java
class Failure {
    interface Box<A> {
        <B extends A> B get();
    }

    <S extends String> Box<S[]> getBox() {
	return null;
    }

    void m() {
        for (String s : getBox().get()) {}
    }
}
```

and you will get:

```
error: SourceChecker.typeProcess: unexpected Throwable (AssertionError) while processing Issue419.java; message: ArrayType must be represented by AnnotatedArrayType
  Compilation unit: Issue419.java
  Exception: java.lang.AssertionError: ArrayType must be represented by AnnotatedArrayType; Stack trace: org.checkerframework.framework.flow.CFCFGBuilder$CFCFGTranslationPhaseOne.createEnhancedForLoopArrayVariable(CFCFGBuilder.java:148)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitEnhancedForLoop(CFGBuilder.java:3646)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitEnhancedForLoop(CFGBuilder.java:1440)
  com.sun.tools.javac.tree.JCTree$JCEnhancedForLoop.accept(JCTree.java:1043)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:3247)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:1440)
  com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:918)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.process(CFGBuilder.java:1554)
  org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.process(CFGBuilder.java:1601)
  org.checkerframework.framework.flow.CFCFGBuilder.run(CFCFGBuilder.java:61)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1138)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1115)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1045)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1422)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:247)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:292)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:163)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:967)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:502)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:185)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
```

This is related to #979 and came up when working through problems with going to Java 9.
If nothing else, it should be another test case.

--- a/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
+++ b/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
@@ -12,6 +12,8 @@ import java.util.Collection;
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
 import org.checkerframework.common.basetype.BaseTypeChecker;
 import org.checkerframework.dataflow.cfg.CFGBuilder;
 import org.checkerframework.dataflow.cfg.ControlFlowGraph;
@@ -19,6 +21,8 @@ import org.checkerframework.dataflow.cfg.UnderlyingAST;
 import org.checkerframework.framework.source.SourceChecker;
 import org.checkerframework.framework.type.AnnotatedTypeFactory;
 import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
 import org.checkerframework.javacutil.ErrorReporter;
 import org.checkerframework.javacutil.TreeUtils;
 
@@ -144,6 +148,16 @@ public class CFCFGBuilder extends CFGBuilder {
             factory.shouldCache = false;
             AnnotatedTypeMirror annotatedArrayType = factory.getAnnotatedType(expression);
             factory.shouldCache = oldShouldCache;
+            if (annotatedArrayType.getKind() == TypeKind.WILDCARD) {
+                if (((AnnotatedWildcardType) annotatedArrayType).isUninferredTypeArgument()) {
+                    TypeMirror type = TreeUtils.typeOf(expression);
+                    AnnotatedArrayType newArrayType =
+                            (AnnotatedArrayType)
+                                    AnnotatedTypeMirror.createType(type, factory, false);
+                    newArrayType.setComponentType(annotatedArrayType);
+                    annotatedArrayType = newArrayType;
+                }
+            }
 
             assert (annotatedArrayType instanceof AnnotatedTypeMirror.AnnotatedArrayType)
                     : "ArrayType must be represented by AnnotatedArrayType";

--- a/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
+++ b/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
@@ -148,15 +148,13 @@ public class CFCFGBuilder extends CFGBuilder {
             factory.shouldCache = false;
             AnnotatedTypeMirror annotatedArrayType = factory.getAnnotatedType(expression);
             factory.shouldCache = oldShouldCache;
-            if (annotatedArrayType.getKind() == TypeKind.WILDCARD) {
-                if (((AnnotatedWildcardType) annotatedArrayType).isUninferredTypeArgument()) {
-                    TypeMirror type = TreeUtils.typeOf(expression);
-                    AnnotatedArrayType newArrayType =
-                            (AnnotatedArrayType)
-                                    AnnotatedTypeMirror.createType(type, factory, false);
-                    newArrayType.setComponentType(annotatedArrayType);
-                    annotatedArrayType = newArrayType;
-                }
+            if (annotatedArrayType.getKind() == TypeKind.WILDCARD
+                    && ((AnnotatedWildcardType) annotatedArrayType).isUninferredTypeArgument()) {
+                TypeMirror type = TreeUtils.typeOf(expression);
+                AnnotatedArrayType newArrayType =
+                        (AnnotatedArrayType) AnnotatedTypeMirror.createType(type, factory, false);
+                newArrayType.setComponentType(annotatedArrayType);
+                annotatedArrayType = newArrayType;
             }
 
             assert (annotatedArrayType instanceof AnnotatedTypeMirror.AnnotatedArrayType)

--- a/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
+++ b/framework/src/org/checkerframework/framework/flow/CFCFGBuilder.java
@@ -154,6 +154,7 @@ public class CFCFGBuilder extends CFGBuilder {
                 AnnotatedArrayType newArrayType =
                         (AnnotatedArrayType) AnnotatedTypeMirror.createType(type, factory, false);
                 newArrayType.setComponentType(annotatedArrayType);
+                newArrayType.addAnnotations(annotatedArrayType.getEffectiveAnnotations());
                 annotatedArrayType = newArrayType;
             }
 

--- a/framework/tests/all-systems/java8inference/Issue1775.java
+++ b/framework/tests/all-systems/java8inference/Issue1775.java
@@ -1,0 +1,17 @@
+// Test case for issue 1775
+// https://github.com/typetools/checker-framework/issues/1775
+
+@SuppressWarnings("") // just check for crashes
+class Issue1775 {
+    interface Box<A> {
+        <B extends A> B get();
+    }
+
+    <S extends String> Box<S[]> getBox() {
+        return null;
+    }
+
+    void m() {
+        for (String s : getBox().get()) {}
+    }
+}

